Documentation/man/man3/AWEJSONScraper.3
.TH "AWEJSONScraper" 3 "Thu Apr 17 2014" "Version 0.1" "AWE Media Center" \" -*- nroff -*-
.ad l
.nh
.SH NAME
AWEJSONScraper \- 
.PP
A metadata scraper based on JSON files\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <AWEJSONScraper\&.h>\fP
.PP
Inherits \fBAWEMetadataScraper\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBAWEJSONScraper\fP (const std::string &name, const std::string &type)"
.br
.RI "\fIConstruct a new JSON-based scraper from the given name\&. \fP"
.ti -1c
.RI "virtual bool \fBprepare\fP (\fBAWEGlobalSettings\fP *settings)"
.br
.RI "\fIPrepares the scraper by reading the scraper and type files\&. \fP"
.ti -1c
.RI "virtual bool \fBscrapeDataForFile\fP (\fBAWEMediaFile\fP *file, bool askUser, bool import)"
.br
.RI "\fIRetrieves metadata for a media file\&. \fP"
.ti -1c
.RI "virtual void \fBdeactivate\fP ()"
.br
.RI "\fIDestroys any used-up dynamic memory\&. \fP"
.ti -1c
.RI "virtual bool \fBisValid\fP ()"
.br
.RI "\fIDetermines if this is scraper can be used\&. \fP"
.ti -1c
.RI "virtual const std::string & \fBgetName\fP ()"
.br
.RI "\fIGets the name of the scraper\&. \fP"
.ti -1c
.RI "virtual const std::string & \fBgetType\fP ()"
.br
.RI "\fIGets the media type for this scraper\&. \fP"
.in -1c
.SH "Detailed Description"
.PP 
A metadata scraper based on JSON files\&. 


.PP
\fBTodo\fP
.RS 4
Change \fCmyType\fP to be the mystical know-all configuration object\&.
.RE
.PP
.PP
One way of defining a scraper is through a JSON file interpreted by an instantiation of this class\&. To read more about the format of such a JSON file, read \fBthis document\fP\&.
.PP
For information on media types, read \&.\&./type/README\&.md 'this document'\&. 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "AWEJSONScraper::AWEJSONScraper (const std::string &name, const std::string &type)"

.PP
Construct a new JSON-based scraper from the given name\&. 
.PP
\fBParameters:\fP
.RS 4
\fIname\fP The name of the scraper\&. 
.br
\fItype\fP The media type for this scraper\&. 
.RE
.PP

.SH "Member Function Documentation"
.PP 
.SS "void AWEJSONScraper::deactivate ()\fC [virtual]\fP"

.PP
Destroys any used-up dynamic memory\&. This helps with memory management\&. 
.PP
Implements \fBAWEMetadataScraper\fP\&.
.SS "const std::string & AWEJSONScraper::getName ()\fC [virtual]\fP"

.PP
Gets the name of the scraper\&. 
.PP
\fBReturns:\fP
.RS 4
The name of the scraper\&. 
.RE
.PP

.PP
Implements \fBAWEMetadataScraper\fP\&.
.SS "const std::string & AWEJSONScraper::getType ()\fC [virtual]\fP"

.PP
Gets the media type for this scraper\&. 
.PP
\fBReturns:\fP
.RS 4
The media type name for the scraper\&. 
.RE
.PP

.PP
Implements \fBAWEMetadataScraper\fP\&.
.SS "bool AWEJSONScraper::isValid ()\fC [virtual]\fP"

.PP
Determines if this is scraper can be used\&. 
.PP
\fBReturns:\fP
.RS 4
\fCtrue\fP if this scraper can be used successfully, \fCfalse\fP otherwise\&. 
.RE
.PP

.PP
Implements \fBAWEMetadataScraper\fP\&.
.SS "bool AWEJSONScraper::prepare (\fBAWEGlobalSettings\fP *settings)\fC [virtual]\fP"

.PP
Prepares the scraper by reading the scraper and type files\&. This helps with memory management\&.
.PP
\fBParameters:\fP
.RS 4
\fIsettings\fP The \&.\&./settings/README\&.md 'global settings of AWEMC'\&.
.RE
.PP
\fBReturns:\fP
.RS 4
\fCtrue\fP if the scraper was able to prepare itself, \fCfalse\fP if an error occured and scraping should be aborted\&. Generally, a \fCfalse\fP result here means that the JSON file was incorrectly written\&. 
.RE
.PP

.PP
Implements \fBAWEMetadataScraper\fP\&.
.SS "bool AWEJSONScraper::scrapeDataForFile (\fBAWEMediaFile\fP *file, boolaskUser, boolimport)\fC [virtual]\fP"

.PP
Retrieves metadata for a media file\&. To construct your JSON file, you should use \fCjsoncpp\fP\&.
.PP
\fCaskUser\fP tells your scraper if the user wants to be given a list of choices for certain things\&. You should NOT ask the user for everything if \fCaskUser\fP is true; only basic things like, 'Which icon do you want to use?'
.PP
\fCimport\fP specifies if the files you get should be copied into AWEMC's folders\&. Do NOT copy the media file\&.
.PP
\fBTodo\fP
.RS 4
\fCimport\fP and \fCaskUser\fP
.RE
.PP
.PP
\fBParameters:\fP
.RS 4
\fIfile\fP The media file to get metadata for\&. 
.br
\fIaskUser\fP \fCtrue\fP if the user wants to be given choices, \fCfalse\fP otherwise\&. 
.br
\fIimport\fP \fCtrue\fP if the user wants to import metadata files, \fCfalse\fP otherwise\&.
.RE
.PP
\fBReturns:\fP
.RS 4
\fCtrue\fP if the scraper was able to get the metadata, \fCfalse\fP if it was not\&. 
.RE
.PP

.PP
Implements \fBAWEMetadataScraper\fP\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for AWE Media Center from the source code\&.


Code/scraper/AWEJSONScraper.cpp
// class header
#include "AWEJSONScraper.h"

// for reading and writing
#include "libs/generic_file_reader/file_reader.h"
#include <sstream>
#include <iostream>
#include <fstream>

// debug
#include "debug_macros/debug.h"

using namespace AWE;
using namespace Json;
using namespace std;

JSONScraper::JSONScraper(const string& name,
		const string& type) :
	myValidity(false),
	myName(name),
	myType(type),
	myCurrentFile(0)
	{ }

bool JSONScraper::prepare(GlobalSettings* settings)
{
	myScraper = settings->getScraperSettingsByName(myName);
	myDefaultProperties = settings->getTypeByName(myType);
	myValidity = checkValidity();
	return myValidity;
}

bool JSONScraper::scrapeDataForFile(MediaFile* file,
	bool askUser, bool import)
{
	// set ans to defaults
	myCurrentFile = file;
	string temp = myCurrentFile->getStringMember("metadata.location");
	myCurrentFile->getMember("metadata") = myDefaultProperties;
	myCurrentFile->getMember("metadata.location") = temp;

	// check validity
	if (!isValid())
	{
		DEBUG(cout << "Scraper \"" << getName() << "\" not valid" << endl;)
		return false;
	}

	// get the backrefs for the filename
	BackrefList backrefs;
	if (!getBackrefs(temp, myScraper["filename"].asString(), backrefs))
	{
		return false;
	}

	// run through the procedures, noting issues
	bool procsSucceeded = true;
	for (unsigned int i = 0; i < myScraper["procedures"].size(); ++ i)
	{
		procsSucceeded = procsSucceeded 
			&& executeProcedure(myScraper["procedures"][i], backrefs,
				askUser, import);
	}

	// reset the file variable
	myCurrentFile = 0;

	// return the success/failure
	return procsSucceeded;
}

bool JSONScraper::executeProcedure(Value& procedure, BackrefList& backrefs,
	bool askUser, bool import)
{
	// get the file to look in and what to look for
	string lookInFile = procedure["look in file"].asString();
	string lookFor = procedure["for"].asString();
	if (!replaceBackrefs(lookInFile, backrefs)
		|| !replaceBackrefs(lookFor, backrefs))
	{
		return false;
	}

	// get the contents of the file
	stringstream fileContents;
	copyFile(lookInFile, fileContents);

	BackrefList list;
	bool called = false;
	bool procsSucceeded = true;
	// repetitive (i.e. array-filling) procedure
	if (procedure["repeat"].asBool())
	{
		// for every match
		string s = fileContents.str();
		while (getBackrefs(s, lookFor, list))
		{
			// do the necessary stuff with this result
			procsSucceeded = procsSucceeded
					&& useMatchForProcedure(procedure, list, askUser, import);
			s = list.suffix().str();
			called = true;
		}
		if (!called)
		{
			return false;
		}
	}
	// not repetitive
	else
	{
		if (!getBackrefs(fileContents.str(), lookFor, list))
		{
			return false;
		}
		procsSucceeded = useMatchForProcedure(procedure, list, askUser, import);
	}

	// return success/failure
	return procsSucceeded;
}

bool JSONScraper::useMatchForProcedure(Value& procedure,
	BackrefList& backrefs, bool askUser, bool import)
{
	bool procsSucceeded = true;

	// apply to the properties in this procedure
	Value::Members mems = procedure.getMemberNames();
	Value::Members::iterator it = mems.begin();
	for (auto prop : mems)
	{
		if (prop != "look in file" && prop != "for" 
			&& prop != "repeat" && prop != "procedures"
			&& prop != "ask user")
		{
			// determine what type it is
			Value valueToSet;
			if (procedure[prop].isString()) // string implies backrefs
			{
				// replace backrefs in the value
				string temp = procedure[prop].asString();
				if (!replaceBackrefs(temp, backrefs))
				{
					return false;
				}
				valueToSet = temp;
			}
			else
			{
				valueToSet = procedure[prop];
			}

			// parse the property name to delve down objects
			Value* toSet = &(myCurrentFile->getMember(
									"metadata." + prop));

			// set the object in the appropriate way
			if (toSet->isArray())
			{
				// append to the array
				toSet->append(valueToSet);
			}
			else // something else
			{
				// replace the current value
				toSet->operator= (valueToSet);
			}
		}
		++ it;
	}

	// apply to the sub procedures
	for (unsigned int j = 0; j < procedure["procedures"].size(); ++ j)
	{
		procsSucceeded = procsSucceeded 
			&& executeProcedure(procedure["procedures"][j],
				backrefs, askUser, import);
	}

	return procsSucceeded;
}

void JSONScraper::deactivate()
{
	myScraper = Value::null;
	myDefaultProperties = Value::null;
}

bool JSONScraper::getBackrefs(const string& search, const string& reg,
	BackrefList& backrefs)
{
	// initialize the regex
	regex r(reg);

	// get the matches (only care about the first)
	return regex_search(search, backrefs, r);
}

bool JSONScraper::replaceBackrefs(string& pseudo_reg, 
	const BackrefList& backrefs)
{
	pseudo_reg = backrefs.format(pseudo_reg);

	// always a success (AFAICT)
	return true;
}

bool JSONScraper::isValid()
{
	return myValidity;
}

bool JSONScraper::checkValidity()
{
	// if true, we assume its good
	if (myValidity)
	{
		return true;
	}
	// doesn't have the necessary tags
	if (!myScraper.isMember("filename") || !myScraper["filename"].isString()
		|| !myScraper.isMember("type") || !myScraper["type"].isString()
		|| !myScraper.isMember("procedures") || !myScraper["procedures"].isArray()
		|| !myScraper.isMember("force copy") || !myScraper["force copy"].isArray()
		|| !myScraper.isMember("copy") || !myScraper["copy"].isArray())
	{
		return false;
	}
	// check each procedure
	for (unsigned int i = 0; i < myScraper["procedures"].size(); ++ i)
	{
		if (!checkProcedureValidity(myScraper["procedures"][i]))
		{
			// bad procedure
			return false;
		}
	}
	// all is well
	return true;
}

bool JSONScraper::checkProcedureValidity(Value& procedure)
{
	// doesn't have the necessary tags
	if (!procedure.isMember("repeat") || !procedure["repeat"].isBool()
		|| !procedure.isMember("ask user") || !procedure["ask user"].isBool()
		|| !procedure.isMember("look in file") || !procedure["look in file"].isString()
		|| !procedure.isMember("for") || !procedure["for"].isString()
		|| !procedure.isMember("procedures") || !procedure["procedures"].isArray())
	{
		return false;
	}
	// check each sub procedure
	for (unsigned int i = 0; i < procedure["procedures"].size(); ++ i)
	{
		if (!checkProcedureValidity(procedure["procedures"][i]))
		{
			// bad procedure
			return false;
		}
	}
	// all is well
	return true;
}

const string& JSONScraper::getName()
{
	return myName;
}

const string& JSONScraper::getType()
{
	return myType;
}